### PR TITLE
PG16 - Add citus_truncate_trigger for Citus foreign tables

### PR DESCRIPTION
**Pull request by naisila**
Relevant PG commit:
https://github.com/postgres/postgres/commit/3b00a94

https://github.com/citusdata/citus/issues/7138

### CODE REVIEW

File: src/test/regress/expected/pg16.out [hunk 1]
```diff
@@ -202,6 +202,61 @@ SELECT create_distributed_table('test_storage', 'a');
...
+
+INSERT INTO foreign_table VALUES (2, 'test_2');
+INSERT INTO foreign_table_test VALUES (3, 'test_3');
+CREATE FUNCTION trigger_func() RETURNS trigger LANGUAGE plpgsql AS $$
+BEGIN
+	RAISE NOTICE 'trigger_func(%) called: action = %, when = %, level = %',
+		TG_ARGV[0], TG_OP, TG_WHEN, TG_LEVEL;
+	RETURN NULL;
+END;$$;
+CREATE TRIGGER trig_stmt_before BEFORE TRUNCATE ON foreign_table
```
Comment (onderkalaci) [inline src/test/regress/expected/pg16.out:241]: I think this test ensures that we have the trigger on the shell table.

It is also possible to create triggers on the shards, see https://github.com/citusdata/citus/blob/main/src/test/regress/sql/distributed_triggers.sql

Do you think is it worth adding such test? I think it'd be at least useful to check it locally before merging. 

ps: As far as I remember we do not allow "distributed foreign tables", but still we can probably have a similar test for citus managed table shards?

**Review by onderkalaci — APPROVED** [review]

File: src/test/regress/expected/pg16.out (hunk 1 above)
Comment (naisila) [reply src/test/regress/expected/pg16.out:353]: Good point - I am actually not able to get successful truncate trigger on the shard. It's not about the trigger though - I believe we have something missing on truncating Citus foreign tables logic. Looking into it.

**Comment by naisila** [comment]
Normally, we support truncating Citus tables through the `citus_truncate_trigger`. However, this trigger hasn't been active so far for Citus foreign tables since foreign tables did not support TRUNCATE trigger (until now with PG16 [Support TRUNCATE triggers on foreign tables. · postgres/postgres@3b00a94 (github.com)](https://github.com/postgres/postgres/commit/3b00a94))

[citus/src/backend/distributed/commands/citus_add_local_table_to_metadata.c at main · citusdata/citus (github.com)](https://github.com/citusdata/citus/blob/main/src/backend/distributed/commands/citus_add_local_table_to_metadata.c#L1482)




However, the Citus foreign table local shard was truncated somehow, even without the citus_truncate_trigger ... I mean truncating Citus foreign tables somehow worked so far, we even have this commit [Support TRUNCATE for foreign tables · citusdata/citus@a1c3580 (github.com)](https://github.com/citusdata/citus/commit/a1c3580c64e72d351c5abbc9ae16bf1f7bbde7ee) and the test is passing.

I am confused as to how the truncate worked so far @onderkalaci

File: src/backend/distributed/commands/citus_add_local_table_to_metadata.c [hunk 2]
```diff
@@ -1478,11 +1478,16 @@ InsertMetadataForCitusLocalTable(Oid citusLocalTableId, uint64 shardId,
 static void
 FinalizeCitusLocalTableCreation(Oid relationId)
 {
+#if PG_VERSION_NUM >= PG_VERSION_16
```
Comment (onderkalaci) [inline src/backend/distributed/commands/citus_add_local_table_to_metadata.c:1481]: 1 line comment would help, saying that PG 16+ supports truncate triggers on fdw

**Review by onderkalaci — APPROVED** [review]

**Pull request merged**